### DISCUSSION TITLE
Using with Apple M1 and rebundle jar files

### DISCUSSION THREAD

**Discussion by Me1G** [tofi86/universalJavaApplicationStub#109] [Q&A] [answered]
Hi

Ok so I will start by saying I am totally out my depth. Just created github acc. I have a client with a relatively unknown Java application but it is really functional for what he needs. It works on Windows and Apple but since the client purchased a new MacBook Pro with M1 architecture, the software has stopped working. I have reached out to the developer and he is no longer supporting the application but has given me the source code. I am a IT Administrator who is starting out with getting into software dev and wanting to help find a work around to this issue. Basically the application won't open and gives below error: 

Process:               JavaApplicationStub [6396]
Path:                  /Applications/appname.app/Contents/MacOS/JavaApplicationStub
Identifier:            com.appname.appname.appname
Version:               ???
Code Type:             X86-64 (Translated)
Parent Process:        ??? [1]
Responsible:           JavaApplicationStub [6396]
User ID:               501

Date/Time:             2021-03-01 20:47:59.864 +1100
OS Version:            macOS 11.2.2 (20D80)
Report Version:        12
Anonymous UUID:        B8AB3F14-36C9-28A0-1215-9FD296765929

Sleep/Wake UUID:       63826CBC-A1E6-48DA-BE94-0F20B48EF0C6

Time Awake Since Boot: 2800 seconds
Time Since Wake:       53 seconds

System Integrity Protection: enabled

Crashed Thread:        0

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    DYLD, [0x1] Library missing

Application Specific Information:
dyld: launch, loading dependent libraries

Dyld Error Message:
  dyld: Using shared cache: ED582735-28E0-3BED-8517-F876F6C67F0B
Library not loaded: /System/Library/PrivateFrameworks/JavaApplicationLauncher.framework/Versions/A/JavaApplicationLauncher
  Referenced from: /Applications/APPNAME/APPNAME.app/Contents/MacOS/JavaApplicationStub
  Reason: image not found

Binary Images:
       0x100a2a000 -        0x100a2afff + (???) <2EF297DB-A500-C676-D2E9-3AF8AA29E602> 
       0x200b40000 -        0x200bdbfff  dyld (832.7.3) <0D4EA85F-7E30-338B-9215-314A5A5539B6> /usr/lib/dyld
    0x7ffdffee1000 -     0x7ffdfff54fff +runtime (203.30) <C98E75A6-BDC8-3D5C-B95B-6422005E96D8> /Library/Apple/*/runtime

Model: MacBookPro17,1, BootROM 6723.81.1, proc 8:4:4 processors, 8 GB, SMC 
Graphics: kHW_AppleM1Item, Apple M1, spdisplays_builtin
Memory Module: lp_ddr4
AirPort: spairport_wireless_card_type_airport_extreme, wl0: Dec 31 2020 21:39:03 version 18.20.222.20.7.8.104 FWID 01-1b2645bc
Bluetooth: Version 8.0.3d9, 3 services, 27 devices, 1 incoming serial ports
Network Service: Wi-Fi, AirPort, en0
USB Device: USB 3.1 Bus
USB Device: USB 3.1 Bus
Thunderbolt Bus: MacBook Pro, Apple Inc.
Thunderbolt Bus: MacBook Pro, Apple Inc.

I installed AZUL's open JDK build for ARMv8 (8u282b08 Zulu:8.52.0.23) and after many hours trying to figure out what may be wrong I have come across the universalJavaApplicationStub and it seems it could solve the issue if I knew how to use it correctly or it was tweaked to work with Apple M1 architecture. As you can see from the  error it is a Intel x86-64 app and I believe it should get translated by Rosetta 2. The problem app did not prompt for Rosetta to be installed but I followed instructions from another blog to manually install. It seems that some issues may lie in the apps info.plist file which I tried to amend to use the universalJavaApplicationStub but still no luck. However, I also tested on a older Apple Mac without M1 chip but on Big Sur and it would not work either so maybe it is not the M1 chip itself and something with Big Sur and Java apps that needed the old native Apple java build installed. The app did work on Catalina. 

Maybe I need to rebundle the App differently or something. Think its was bundled for Apple with a program called Packager. 

Any help or advise would be greatly appreciated. Maybe this is the wrong discussion topic but the universalJavaApplicationStub seemed to be the closest I could find to alluded to the problem could be. 

Thank you

**Comment by tofi86** [comment]
Hi,

it's very hard to solve this issue without having access to the actual app code.

First of all, you should replace the `appname.app/Contents/MacOS/JavaApplicationStub` with the `universalJavaApplicationStub` file taken from this repo's `src/` directory. Make sure to make the file executable.

Then, make sure to change the executable name in `appname.app/Contents/Info.plist` from `JavaApplicationStub` to `universalJavaApplicationStub` — or rename the `universalJavaApplicationStub` to `JavaApplicationStub`.

This project aims to support even the old Apple and Oracle Java Plist keys, so it should work without big tweaks to the `Info.plist` file.

If you need further help, post the content of the `Info.plist` file here as plain text, marking the beginning and end with three backticks ```.

**Reply by Me1G** [reply]
Hi

Yes the steps you describe is what I done originally except I had not tried to rename the stub exe to `JavaApplicationStub` which I did now. 

I have the old native Apple Java installed and the Azul JDK for M1 architecture. Do I need Oracle JDK? They don't seem to have build for M1 chip or does that not matter?I was getting same error with macOS Big Sur on old intel Mac also. However app worked  fine on macOS Catalina. So don't even know if the M1 chip is the problem

I now also tried to just compile the basic HelloWorld java app and get same issue saying 

```$ javac HelloWorld.java
The operation couldn’t be completed. Unable to locate a Java Runtime that supports javac.
``` 
so maybe it is not a problem with my app and it is something else. 

The `Info.plist` is below and after that also the error I get if I double click the stub itself. Also after making change I don't get the error from my original post and if I try open  app, it just bounces a few times and nothing happens.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>CFBundleName</key>
	<string>TherapistPro</string>
	<key>CFBundleIdentifier</key>
	<string>com.longgrasssystems.therapistpro.TherapistPro4</string>
	<key>CFBundleVersion</key>
	<string>4.1.0</string>
	<key>CFBundleAllowMixedLocalizations</key>
	<string>true</string>
	<key>CFBundleExecutable</key>
	<string>JavaApplicationStub</string>
	<key>CFBundleDevelopmentRegion</key>
	<string>English</string>
	<key>CFBundlePackageType</key>
	<string>APPL</string>
	<key>CFBundleSignature</key>
	<string>????</string>
	<key>CFBundleInfoDictionaryVersion</key>
	<string>6.0</string>
	<key>CFBundleIconFile</key>
	<string>TherapistProMacIcon.icns</string>
	<key>Java</key>
	<dict>
		<key>VMOptions</key>
		<string>-cp=.:$APP_PACKAGE/lib</string>
		<key>MainClass</key>
		<string>com.longgrasssystems.therapistpro.TherapistPro4</string>
		<key>JVMVersion</key>
		<string>1.6+</string>
		<key>ClassPath</key>
		<array>
			<string>$JAVAROOT/../../../../lib/TherapistProMain.jar</string>
			<string>$JAVAROOT/../../../../lib/swing-layout-1.0.4.jar</string>
			<string>$JAVAROOT/../../../../lib/derby.jar</string>
			<string>$JAVAROOT/../../../../lib/derbyclient.jar</string>
			<string>$JAVAROOT/../../../../lib/derbynet.jar</string>
			<string>$JAVAROOT/../../../../lib/derbytools.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProAppointments.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProClientCare.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProConditions.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProFixLevel.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProHerbs.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProInvoicing.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProPrescriptions.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProShared.jar</string>
			<string>$JAVAROOT/../../../../lib/TherapistProStock.jar</string>
			<string>$JAVAROOT/../../../../lib/Utilities.jar</string>
		</array>
		<key>Properties</key>
		<dict>
			<key>apple.laf.useScreenMenuBar</key>
			<string>true</string>
		</dict>
	</dict>
</dict>
</plist>
```
When I just double click the new stub I get below:

```M-MacBook-Pro:~ m$ /Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS/JavaApplicationStub ; exit;
Unrecognized option: -cp=.:/Applications/TherapistPro4/TherapistPro4.app/lib
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]
```

**Comment by tofi86** [comment, marked as answer]
Try to replace `<key>Java</key>` with `<key>JavaX</key>` and see if it works now…

The old Apple definitely doesn’t work on M1.

The universalJavaApplicationStub doesn’t care what Java version you use. You can see what Java versions are found and selected for execution by the stub if you open `Console.app` and filter for `syslog`. The name of your app should appear with quite a bunch of log messages by the stub...
You could copy & paste them here for further help of it doesn’t work...

**Reply by Me1G** [reply]
Ok done this and below is syslog when I search with app name:

```
Mar  6 09:36:17 Marks-MacBook-Pro com.apple.xpc.launchd[1]: Coalition Cache Hit:
 app<application.com.longgrasssystems.therapistpro.TherapistPro4.878782.1753554(501)> [2803]
Mar  6 09:36:17 Marks-MacBook-Pro com.apple.xpc.launchd[1] \
(application.com.longgrasssystems.therapistpro.TherapistPro4.878782.1753554[33308]): Service exited with abnormal code:
13
```
There is no logs that mention java itself
Also the App used to appear in the dock before but now it disappears immediately after opening

If I double click the stub itself it gives below which is different from before:

```
Marks-MacBook-Pro:~ m$ /Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS/JavaApplicationStub ; exit;
/Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS/JavaApplicationStub: eval: line 280: unexpected EOF while looking for matching `''
/Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS/JavaApplicationStub: eval: line 281: syntax error: unexpected end of file
xargs: unterminated quote
xargs: unterminated quote
/Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS/JavaApplicationStub: line 309: [: too many arguments
/Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS/JavaApplicationStub: line 865: cd: /Applications/TherapistPro4/TherapistPro4.app/Contents/Java: No such file or directory
logout
Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.
```

**Reply by tofi86** [reply]
you need to search/filter for `syslog` in Console.app, not for your app name. That’s a bit confusing... have a look at the screenshot in README.md

I’ll look into the error stack later when I have enough log information...

**Reply by Me1G** [reply]
Hi

See below. the Syslog after trying to open app 2 times

```
default	01:04:47.313528+1100	syslog	[3753][JavaApplicationStub] [StubDir] /Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS
default	01:04:47.331777+1100	syslog	[3753][TherapistPro] [PlistStyle] Apple
default	01:04:47.370018+1100	syslog	[3753][TherapistPro] [JavaRequirement] JVM minimum version: 1.6+
default	01:04:47.372853+1100	syslog	[3753][TherapistPro] [JavaRequirement] JVM maximum version:
default	01:04:47.375778+1100	syslog	[3753][TherapistPro] [LanguageSearch] Checking preferred languages in macOS System Preferences...
default	01:04:47.385049+1100	syslog	[3753][TherapistPro] [LanguageSearch] ... found [en-AU]
default	01:04:47.387376+1100	syslog	[3753][TherapistPro] [LanguageSearch] ... selected 'en-AU' ('en') as the default language for the launcher stub
default	01:04:47.389679+1100	syslog	[3753][TherapistPro] [Language] en
default	01:04:47.395982+1100	syslog	[3753][TherapistPro] [JavaSearch] Checking for $JAVA_HOME ...
default	01:04:47.398576+1100	syslog	[3753][TherapistPro] [JavaSearch] ... haven't found JAVA_HOME
default	01:04:47.401163+1100	syslog	[3753][TherapistPro] [JavaSearch] Searching for JavaVirtualMachines on the system ...
default	01:04:47.416291+1100	syslog	[3753][TherapistPro] [JavaSearch] ... found JVM: 1.8.0_282:/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java
default	01:04:47.418766+1100	syslog	[3753][TherapistPro] [JavaSearch] Filtering the result list for JVMs matching the min/max version requirement ...
default	01:04:47.428519+1100	syslog	[3753][TherapistPro] [JavaSearch] ... matches all requirements: 08000282:1.8.0_282:/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java
default	01:04:47.432334+1100	syslog	[3753][TherapistPro] [JavaCommand] '/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java'
default	01:04:47.469122+1100	syslog	[3753][TherapistPro] [JavaVersion] 1.8.0_282 / 08000282
default	01:04:47.472318+1100	syslog	[3753][TherapistPro] [WorkingDirectory] /Applications/TherapistPro4
default	01:04:47.475704+1100	syslog	[3753][TherapistPro] [Exec] "/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java" -cp ".:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProMain.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/swing-layout-1.0.4.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derby.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbyclient.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbynet.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbytools.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProAppointments.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProClientCare.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Jav<…>
default	01:09:13.997148+1100	syslog	[4348][JavaApplicationStub] [StubDir] /Applications/TherapistPro4/TherapistPro4.app/Contents/MacOS
default	01:09:14.016891+1100	syslog	[4348][TherapistPro] [PlistStyle] Apple
default	01:09:14.053714+1100	syslog	[4348][TherapistPro] [JavaRequirement] JVM minimum version: 1.6+
default	01:09:14.056662+1100	syslog	[4348][TherapistPro] [JavaRequirement] JVM maximum version:
default	01:09:14.060018+1100	syslog	[4348][TherapistPro] [LanguageSearch] Checking preferred languages in macOS System Preferences...
default	01:09:14.071429+1100	syslog	[4348][TherapistPro] [LanguageSearch] ... found [en-AU]
default	01:09:14.074331+1100	syslog	[4348][TherapistPro] [LanguageSearch] ... selected 'en-AU' ('en') as the default language for the launcher stub
default	01:09:14.076627+1100	syslog	[4348][TherapistPro] [Language] en
default	01:09:14.082387+1100	syslog	[4348][TherapistPro] [JavaSearch] Checking for $JAVA_HOME ...
default	01:09:14.084739+1100	syslog	[4348][TherapistPro] [JavaSearch] ... haven't found JAVA_HOME
default	01:09:14.087110+1100	syslog	[4348][TherapistPro] [JavaSearch] Searching for JavaVirtualMachines on the system ...
default	01:09:14.102745+1100	syslog	[4348][TherapistPro] [JavaSearch] ... found JVM: 1.8.0_282:/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java
default	01:09:14.105571+1100	syslog	[4348][TherapistPro] [JavaSearch] Filtering the result list for JVMs matching the min/max version requirement ...
default	01:09:14.114846+1100	syslog	[4348][TherapistPro] [JavaSearch] ... matches all requirements: 08000282:1.8.0_282:/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java
default	01:09:14.119033+1100	syslog	[4348][TherapistPro] [JavaCommand] '/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java'
default	01:09:14.178216+1100	syslog	[4348][TherapistPro] [JavaVersion] 1.8.0_282 / 08000282
default	01:09:14.180794+1100	syslog	[4348][TherapistPro] [WorkingDirectory] /Applications/TherapistPro4
default	01:09:14.183650+1100	syslog	[4348][TherapistPro] [Exec] "/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java" -cp ".:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProMain.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/swing-layout-1.0.4.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derby.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbyclient.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbynet.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbytools.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProAppointments.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProClientCare.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Jav<…>
```

**Reply by tofi86** [reply]
I think the problem is related to the path of your libs. Have a look at the executed command:

`[Exec] "/Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home/bin/java" -cp ".:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProMain.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/swing-layout-1.0.4.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derby.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbyclient.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbynet.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/derbytools.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProAppointments.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/../../../../lib/TherapistProClientCare.jar:/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Jav<…>`

The launching App tries to access JAR files from a `lib/` folder _relative_ to the App itself. Which in this case seems to be in `/Applications/TherapistPro4/lib/`, next to the `TherapistPro4.app`, right?

The problem is, that in latest macOS versions, apps are executed in sandboxed environments and cannot access resources outside the sandboxed app itself. Or at least not per default.

Please test the following:

1. Does `/Applications/TherapistPro4/lib/` exist and does it contain the mentioned JAR files?
2. If so, it's most likely a sandbox issue.
3. If so, move the complete `lib/` folder _inside_ the app to `/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/` (known as `$JAVAROOT`) and change the `Info.plist` file by replacing every occurance of `$JAVAROOT/../../../../lib/` with `$JAVAROOT/lib/`.
4. If that doesn't work, try to copy the WHOLE value of the `[Exec]` log line in a comment here.

**Reply by Me1G** [reply]
I tried this and it still didn't work. Not sure what you mean by the whole ```[Exec]``` 

Assuming you have an M1 Mac you could download the program from [http://www.birdsongsoftware.com.au/therapistprodownloads.html]()

That way you could test better or see the folder and package contents

Also I tried to change the ```
<string>-cp=.:$APP_PACKAGE/lib</string> ``` as I thought this was maybe not pointing to inside the bundle

**Reply by tofi86** [reply]
This altered `Info.plist` works for me:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>CFBundleName</key>
	<string>TherapistPro</string>
	<key>CFBundleIdentifier</key>
	<string>com.longgrasssystems.therapistpro.TherapistPro4</string>
	<key>CFBundleVersion</key>
	<string>4.1.0</string>
	<key>CFBundleAllowMixedLocalizations</key>
	<string>true</string>
	<key>CFBundleExecutable</key>
	<string>universalJavaApplicationStub</string>
	<key>CFBundleDevelopmentRegion</key>
	<string>English</string>
	<key>CFBundlePackageType</key>
	<string>APPL</string>
	<key>CFBundleSignature</key>
	<string>????</string>
	<key>CFBundleInfoDictionaryVersion</key>
	<string>6.0</string>
	<key>CFBundleIconFile</key>
	<string>TherapistProMacIcon.icns</string>
	<key>JavaX</key>
	<dict>
		<key>MainClass</key>
		<string>com.longgrasssystems.therapistpro.TherapistPro4</string>
		<key>JVMVersion</key>
		<string>1.5+</string>
		<key>ClassPath</key>
		<array>
			<string>$APP_PACKAGE/../lib/TherapistProMain.jar</string>
			<string>$APP_PACKAGE/../lib/swing-layout-1.0.4.jar</string>
			<string>$APP_PACKAGE/../lib/derby.jar</string>
			<string>$APP_PACKAGE/../lib/derbyclient.jar</string>
			<string>$APP_PACKAGE/../lib/derbynet.jar</string>
			<string>$APP_PACKAGE/../lib/derbytools.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProAppointments.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProClientCare.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProConditions.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProFixLevel.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProHerbs.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProInvoicing.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProPrescriptions.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProShared.jar</string>
			<string>$APP_PACKAGE/../lib/TherapistProStock.jar</string>
			<string>$APP_PACKAGE/../lib/Utilities.jar</string>
		</array>
		<key>Properties</key>
		<dict>
			<key>apple.laf.useScreenMenuBar</key>
			<string>true</string>
		</dict>
	</dict>
</dict>
</plist>
```

However, I still get an error on startup:

<img width="602" alt="Bildschirmfoto 2021-03-13 um 14 32 56" src="https://user-images.githubusercontent.com/365118/111031704-20839b00-8409-11eb-9172-a8bea5e89aad.png">

I don't think that this is related to the launcher stub. I'd guess it's an issue in the Therapist Pro app itself...

**Reply by Me1G** [reply]
Excellent thanks. I was able to open even though after latest MacOS update I then could not open unless I executed via the stub rather than double clicking the app.

At least now I can move forward. Thanks for all your help

**Reply by tofi86** [reply]
Glad to hear it works for you as well! :-)

**Reply by Me1G** [reply]
Hi So what is the difference between the ``` string>$APP_PACKAGE```
and ```$JAVAROOT```

**Reply by tofi86** [reply]
`$APP_PACKAGE` resolves to `/Applications/TherapistPro4/TherapistPro4.app/` while `$JAVAROOT` resolves to `/Applications/TherapistPro4/TherapistPro4.app/Contents/Resources/Java/` as you can also see in the logs.

However, changing these vars and the relative paths isn't necessary. The real impact had removing the

```
		<key>VMOptions</key>
		<string>-cp=.:$APP_PACKAGE/lib</string>
```

because its already defined by

```
		<key>ClassPath</key>
		<array>
			<string>$APP_PACKAGE/../lib/TherapistProMain.jar</string>
            ...
```

which results in the `-cp` parameter which cannot be set a second time.

See the readme for the generated params: https://github.com/tofi86/universalJavaApplicationStub/#how-the-script-works
